File: 139-word-break/word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        setWordDict = set(wordDict)
        memo = {}
        def checkWordInDict(word: str):
            if word in memo:
                return memo[word]
            if not word:
                return True
            subStr = ""
            for i in range(0, len(word)):
                subStr += word[i]
                if subStr in setWordDict:
                    memo[subStr] = True
                    if (i+1 == len(word)):
                        return True
                    if (checkWordInDict(word[i+1:])):
                        return True
                else:
                    memo[word] = False

            return False
        return checkWordInDict(s)

File: 139-word-break/word-break.java
class Solution {
    public boolean checkWordInDictionary(String word, Map<String, Boolean> visited, Set<String> wordSet)    {
        if (word.isEmpty()) {
            return true;
        }
        if (visited.containsKey(word))  {
            return visited.get(word);
        }
        StringBuilder subWord = new StringBuilder();
        for (int i = 0; i < word.length(); i++)   {
            subWord.append(word.charAt(i));
            if (wordSet.contains(subWord.toString()))    {
                visited.put(subWord.toString(), true);
                if (i+1 == word.length()) {
                    return true;
                }
                if (checkWordInDictionary(word.substring(i+1), visited, wordSet))  {
                    return true;
                }
            }
            else    {
                visited.put(subWord.toString(), false);
            }
        }
        return false;
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        Map<String, Boolean> visited = new HashMap<String, Boolean>();
        Set<String> wordSet = new HashSet<String>(wordDict);
        return checkWordInDictionary(s, visited, wordSet);
    }
}